index.ts

```diff
@@ -1,61 +1,61 @@
+    return path.reduce((current, next) => {
+        if (+next === +next) {
+            current += `[${next}]`;
+        } else {
+            current += current === '' ? `${next}` : `.${next}`;
+        }
 
-export type TypedPathNode<T> = {
-    $path: string;
-    $raw: TypedPathKey[];
-};
+        return current;
+    }, '');
+}
+
+export type TypedPathKey = string | symbol | number;
 
 export type TypedPathFunction<T> = (...args: any[]) => T;
 
-export type TypedPathWrapper<T> = (T extends Array<infer Z>
+export type TypedPathHandlersConfig = Record<string, <T extends TypedPathHandlersConfig = Record<never, never>>(path: string[], additionalHandlers?: T) => any>;
+
+const defaultHandlersConfig = {
+    $path: (path: string[]) => pathToString(path),
+    $raw: (path: string[]) => path,
+    toString: (path: string[]) => () => pathToString(path),
+    [Symbol.toStringTag]: (path: string[]) => () => pathToString(path),
+    valueOf: (path: string[]) => () => pathToString(path),
+}
+
+export type TypedPathHandlers<T extends TypedPathHandlersConfig> = {
+    [key in keyof T]: ReturnType<T[key]>;
+};
+
+export type TypedPathWrapper<T, TPH extends TypedPathHandlers<Record<never, never>>> = (T extends Array<infer Z>
     ? {
-        [index: number]: TypedPathWrapper<Z>;
+        [index: number]: TypedPathWrapper<Z, TPH>;
     }
     : T extends TypedPathFunction<infer RET>
         ? {
-            (): TypedPathWrapper<RET>;
+            (): TypedPathWrapper<RET, TPH>;
         } & {
-            [P in keyof RET]: TypedPathWrapper<RET[P]>;
+            [P in keyof RET]: TypedPathWrapper<RET[P], TPH>;
         }
         : {
-            [P in keyof T]: TypedPathWrapper<T[P]>;
+            [P in keyof T]: TypedPathWrapper<T[P], TPH>;
         }
-    ) & TypedPathNode<T>;
-
-const toStringMethods: (string | symbol | number)[] = [
-    'toString',
-    Symbol.toStringTag,
-    'valueOf'
-];
+    ) & TypedPathHandlers<TPH>;
 
-function pathToString(path: string[]): string {
-    return path.reduce((current, next) => {
-        if (+next === +next) {
-            current += `[${next}]`;
-        } else {
-            current += current === '' ? `${next}` : `.${next}`;
-        }
-
-        return current;
-    }, '');
-}
-
-export function typedPath<T>(path: string[] = []): TypedPathWrapper<T> {
-    return <TypedPathWrapper<T>>new Proxy({}, {
+export function typedPath<T, K extends TypedPathHandlersConfig = Record<never, never>>(path: string[] = [], additionalHandlers?: K): TypedPathWrapper<T, K & typeof defaultHandlersConfig> {
+    return <TypedPathWrapper<T, K & typeof defaultHandlersConfig>>new Proxy({}, {
         get(target: T, name: TypedPathKey) {
-            if (name === '$path') {
-                return pathToString(path);
-            }
-
-            if (name === '$raw') {
-                return path;
-            }
+            const handlersConfig: TypedPathHandlersConfig = { ...(additionalHandlers ?? {}), ...defaultHandlersConfig };
```

Comment:
Missed this comment the first time, sorry.
My concern with caching here is that we can never know what is passed in the `additionalHandlers` object, so we can't be sure that a user won't use a reserved name that we might be using to perform this check. I tried restricting allowed keys for `additionalHandlers` with TypeScript but unfortunately it's not possible at the moment.
Also, I'm not too familiar with the complexity notation, but I just don't think it should matter in this case much because the variable would be garbage collected anyway, right? So basically we're sacrificing resources for the sake of robustness here. Not sure what's the right approach.